crates/uv/src/commands/auth/helper.rs

```diff
@@ -0,0 +1,143 @@
+use uv_preview::Preview;
+use uv_redacted::DisplaySafeUrl;
+
+use crate::{commands::ExitStatus, printer::Printer, settings::NetworkSettings};
+
+/// Request format for the Bazel credential helper protocol.
+#[derive(Debug, Deserialize)]
+struct BazelCredentialRequest {
+    uri: DisplaySafeUrl,
+}
+
+impl BazelCredentialRequest {
+    fn from_str(s: &str) -> Result<Self> {
+        serde_json::from_str(s).context("Failed to parse credential request as JSON")
+    }
+
+    fn from_stdin() -> Result<Self> {
+        let mut buffer = String::new();
+        std::io::stdin()
+            .read_to_string(&mut buffer)
+            .context("Failed to read from stdin")?;
+
+        Self::from_str(&buffer)
+    }
+}
+
+/// Response format for the Bazel credential helper protocol.
+#[derive(Debug, Serialize, Default)]
+struct BazelCredentialResponse {
+    headers: HashMap<String, Vec<String>>,
+}
+
+impl TryFrom<Credentials> for BazelCredentialResponse {
+    fn try_from(creds: Credentials) -> Result<Self> {
+        let header_str = creds
+            .to_header_value()
+            .to_str()
+            // TODO: this is infallible in practice
+            .context("Failed to convert header value to string")?
+            .to_owned();
+
+        Ok(Self {
+            headers: HashMap::from([("Authorization".to_owned(), vec![header_str])]),
+        })
+    }
+
+    type Error = anyhow::Error;
+}
+
+async fn credentials_for_url(
+    url: &DisplaySafeUrl,
+    preview: Preview,
+    network_settings: &NetworkSettings,
+) -> Result<Option<Credentials>> {
+    let pyx_store = PyxTokenStore::from_settings()?;
+
+    // Use only the username from the URL, if present - discarding the password
+    let url_credentials = Credentials::from_url(url);
+    let username = url_credentials.as_ref().and_then(|c| c.username());
+    if url_credentials
+        .as_ref()
+        .map(|c| c.password().is_some())
+        .unwrap_or(false)
+    {
+        debug!("URL '{url}' contain a password; ignoring");
+    }
+
+    if pyx_store.is_known_domain(url) {
+        if username.is_some() {
+            bail!(
+                "Cannot specify a username for URLs under {}",
+                url.host()
+                    .map(|host| host.to_string())
+                    .unwrap_or(url.to_string())
+            );
+        }
+        let client = BaseClientBuilder::new(
+            network_settings.connectivity,
+            network_settings.native_tls,
+            network_settings.allow_insecure_host.clone(),
+            preview,
+            network_settings.timeout,
+            network_settings.retries,
+        )
+        .auth_integration(uv_client::AuthIntegration::NoAuthMiddleware)
+        .build();
+        let maybe_token = pyx_store
+            .access_token(client.for_host(pyx_store.api()).raw_client(), 0)
+            .await
+            .context("Authentication failure")?;
+        let token = maybe_token.context("No access token found")?;
```

Comment:
nit: In general, I'd avoid the intermediate variable? I think `maybe_` variables are weird in Rust.

```suggestion
        let token = pyx_store
            .access_token(client.for_host(pyx_store.api()).raw_client(), 0)
            .await
            .context("Authentication failure")?
            .context("No access token found")?;
```